### PR TITLE
port raftstore SnapManager

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@zhangjinpeng1987 PTAL

File: tikv/raftstore/snap_manager.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+	} else if _, ok := sm.registry[key]; ok {
+		log.Infof("skip to delete %s since it's registered.", snapshot.Path())
+		return false
+	}
+	snapshot.Delete()
+	return true
+}
+
+type SnapManagerBuilder struct {
+	maxTotalSize        uint64
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/snap_manager.go]: The max_write_bytes_per_sec is missing?

Comment (coocood) [reply tikv/raftstore/snap_manager.go]: badger doesn't support the limiter feature yet.

File: tikv/raftstore/snap_manager.go [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	limiter      *IOLimiter
+	MaxTotalSize uint64
+}
+
+func NewSnapManager(path string, ch chan<- *Msg) *SnapManager {
+	return new(SnapManagerBuilder).Build(path, ch)
+}
+
+func (sm *SnapManager) init() error {
+	fi, err := os.Stat(sm.base)
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/snap_manager.go]: Should acquire write lock to prevent multiple threads call init at the same time.

Comment (coocood) [reply tikv/raftstore/snap_manager.go]: I suppose SnapManger.init() would only be called once.

File: tikv/raftstore/snap_manager.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+		}
+		return keyI.RegionID < keyJ.RegionID
+	})
+	return results, nil
+}
+
+func (sm *SnapManager) HasRegistered(key SnapKey) bool {
+	sm.registryLock.RLock()
+	_, ok := sm.registry[key]
+	sm.registryLock.RUnlock()
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/snap_manager.go]: Why not use defer?

File: tikv/raftstore/snap_manager.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+	return ok
+}
+
+func (sm *SnapManager) GetTotalSnapSize() uint64 {
+	return uint64(atomic.LoadInt64(sm.snapSize))
+}
+
+func (sm *SnapManager) GetSnapshotForBuilding(key SnapKey, bundle *DBBundle) (Snapshot, error) {
+	if sm.GetTotalSnapSize() > sm.MaxTotalSize {
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/snap_manager.go]: s/if/while

File: tikv/raftstore/snap_manager.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			return err
+		}
+		snaps = append(snaps, snapWithModTime{key: idleSnap.SnapKey, snap: snap, modTime: fi.ModTime()})
+	}
+	sort.Slice(snaps, func(i, j int) bool {
+		return snaps[i].modTime.Before(snaps[j].modTime)
+	})
+	for sm.GetTotalSnapSize() > sm.MaxTotalSize {
+		if len(snaps) == 0 {
+			return errors.New("too many snapshots")
```
Comment (zhangjinpeng1987) [inline tikv/raftstore/snap_manager.go]: Should collect several rounds?

Comment (zhangjinpeng1987) [reply tikv/raftstore/snap_manager.go]: There maybe some old snapshots can be deleted in the next round.

File: tikv/raftstore/snap_manager.go (hunk 2 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/snap_manager.go]: OK

File: tikv/raftstore/snap_manager.go (hunk 1 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/snap_manager.go]: Got it.

File: tikv/raftstore/snap_manager.go (hunk 3 above)
Comment (coocood) [reply tikv/raftstore/snap_manager.go]: I think It's ok to not defer if there is just one line of code.

File: tikv/raftstore/snap_manager.go (hunk 4 above)
Comment (coocood) [reply tikv/raftstore/snap_manager.go]: `for` loop is done in `deleteOldIdleSnaps`.
Here we just need `if`.

File: tikv/raftstore/snap_manager.go (hunk 5 above)
Comment (coocood) [reply tikv/raftstore/snap_manager.go]: The passed duration is very short, it's unlikely that `ListIdleSnap` again can avoid this error.
And the code is simpler when we only call `ListIdleSnap` once.

**Pull request merged**